### PR TITLE
Issue 4972 - Milestone 1 Drawing Endpoints

### LINKED ISSUE #4972: [2D MVP] Milestone 1: Ability to add/remove drawing containers
### Description
This is part of https://github.com/3drepo/3D-Repo-Product-Team/issues/377.

Create endpoints to provide general drawing container management (note: permissions will be a separate issue). This issue aims to allow the user to create a drawing container, edit its properties and delete them.

##### Drawing categories
  
endpoint: `GET /teamspaces/{teamspace}/projects/{project}/settings/drawingCategories`
Description: Returns the list of drawing categories available within the project
Permissions required: Teamspace/project admin
Response: 
````js
{
  drawingCategories: [
    "a",
    "b"
  ]
}
````

NOTE:
At this stage we're not storing a user editable list of drawing categories. so make it return the hard coded list of model categories (from container)

##### Create a drawing container
endpoint: `POST /teamspaces/{teamspace}/projects/{project}/drawings`
Description: Creates a new drawing container
Permissions required: Teamspace/project admin
Payload:
````js
{
  name: "Level 2 Floor plan", //Name of the drawing, unique within project (strings.title)
  number: "SC1-SFT-V1-01-M3-ST-30_10_30-0001", //Some sort of unique identifier for the drawing (string.title - alphanumeric with underscores and hyphens)
  type: "Architecture",
  desc: "This is the level 2 floor plan for a building in xyz" // strings.shortDescription
}
````

Respsonse:
HTTP 200
````js
{ 
  _id: "3f6eeebd-6f2a-443d-91c1-ef208265771b"  //ID of the drawing that was created
}

````

NOTES:
- In general we want to treat drawings like a different type of container
- This should create an entry in `settings` collection, specifying the type of the model to be `drawing`
- A socket io event should trigger for the new container

##### Edit a drawing container's settings
endpoint: `PATCH /teamspaces/{teamspace}/projects/{project}/drawings/{drawing id}`
Description: Edits a drawing settings
Permissions required: Teamspace/project admin
Payload: Any of the fields that are filled in when the drawing is created

NOTE: 
- A socket io event should trigger for settings being updated

##### Delete a drawing container
endpoint: `DELETE /teamspaces/{teamspace}/projects/{project}/drawings/{drawing id}`
Description: Deletes a drawing container
Permissions required: Teamspace/project admin

NOTE: 
- A socket io event should trigger for container removed

##### Adding favourate drawings
endpoint: `PATCH /teamspaces/{teamspace}/projects/{project}/drawings/favourites`
Description: Adds to the list of favourite drawings within the project
Payload:
````js
{
  models: [
    "3f6eeebd-6f2a-443d-91c1-ef208265771b",
    "ee7e2c5c-a9f0-4a50-9a49-c6daec0d6f7e"
  ]
}
````

##### Removing favourate drawings
endpoint: `DELETE/teamspaces/{teamspace}/projects/{project}/drawings/favourites?ids={drawing id}`
Description: remove from the list of favourite drawings within the project

### Goals
- As an admin, I want to be able to create drawing entries so collaborators and upload their files.
- As an admin I want to specify properties of the drawing so they can be easily identified.
- As an admin, I want to be able to edit these properties
- As an admin, I want to be able to remove a drawing that is no longer applicable
- As an admin, I want to be told if I am trying to create a new drawing container with an existing drawing name
- As an admin, I want to be told if I am trying to create a new drawing container with an existing drawing number
- As a user, I want to be able, to declare a drawing as a favourite/unfavourite

### Tasks
Get Drawing Categories 
- [x] Create route method and swagger doc
- [x] Create processor method
- [ ] Create unit and e2e tests

Create Drawing Container
- [x] Create route method and swagger doc
- [x] Create yup validation method
- [x] Create processor method
- [x] Create model method
- [ ] Create unit and e2e tests

Edit Drawing Container
- [x] Create route method and swagger doc
- [x] Create yup validation method
- [x] Create processor method
- [x] Create model method
- [ ] Create unit and e2e tests

Delete Drawing Container
- [x] Create route method and swagger doc
- [x] Create yup validation method
- [x] Create processor method
- [x] Create model method
- [ ] Create unit and e2e tests

Add Favourite Drawing
- [x] Create route method and swagger doc
- [x] Create yup validation method
- [x] Create processor method
- [x] Create model method
- [ ] Create unit and e2e tests

Remove Favourite Drawing
- [x] Create route method and swagger doc
- [x] Create yup validation method
- [x] Create processor method
- [x] Create model method
- [ ] Create unit and e2e tests

### Related Resources
Product issue: https://github.com/3drepo/3D-Repo-Product-Team/issues/377

### PR DESCRIPTION
**Pull request by ChristosTsiotsias**
This fixes #4972

#### Description
Enpoint for Getting drawing categories
Enpoints for Adding / Updating / Deleting drawings added
Enpoints for Appending / Deleting favourite drawings
Unit & e2e tests added
Swagger documentation added

#### Test cases
All endpoints should function as expected

### CODE REVIEW

File: backend/src/v5/routes/teamspaces/projects/models/common/general.js [hunk 1]
```diff
@@ -20,23 +20,32 @@ const { formatModelSettings, formatModelStats } = require('../../../../../middle
...
 const { getUserFromSession } = require('../../../../../utils/sessions');
 const { respond } = require('../../../../../utils/responder');
 
-const addModel = (isFed) => async (req, res) => {
+const addModel = (modelType) => async (req, res) => {
 	const { teamspace, project } = req.params;
 	try {
-		const fn = isFed ? Federations.addFederation : Containers.addContainer;
+		let fn;
+		if (modelType === MODEL_TYPES.federation) {
```
[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/src/v5/routes/teamspaces/projects/models/common/general.js:43]: similar for the changes here in the rest of the file 😄

File: backend/src/v5/routes/teamspaces/projects/models/common/general.js [hunk 2]
```diff
@@ -140,11 +186,21 @@ const getModelSettings = (isFed) => async (req, res, next) => {
...
+
+	let hasAdminAccessToModel;
+	if (modelType === MODEL_TYPES.federation) {
+		hasAdminAccessToModel = hasAdminAccessToFederation;
+	} else {
+		hasAdminAccessToModel = modelType === MODEL_TYPES.drawing
+			? hasAdminAccessToDrawing : hasAdminAccessToContainer;
+	}
+
+	const hasReadAccessToModel = modelType === MODEL_TYPES.federation
```
[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/src/v5/routes/teamspaces/projects/models/common/general.js:200]: sorry I might've missed it in the scoping but there should be an endpoint to get drawing settings for this issue.

**Review by carmenfan — CHANGES_REQUESTED** [review]
chose the wrong option last time!

See comments above

File: backend/tests/v5/unit/middleware/dataConverter/inputs/teamspaces/projects/models/commons/modelSettings.test.js [hunk 3]
```diff
@@ -53,62 +55,71 @@ Legends.checkLegendExists.mockImplementation((teamspace, model, legend) => {
...
+	if ((modelType === MODEL_TYPES.DRAWING && existingModelNumber.match(number)) && _id?.$in?.length) return {};
 	if (existingModelName.match(name) && _id?.$in?.length) return {};
+
 	throw templates.modelNotFound;
 });
 
 Projects.getProjectById.mockResolvedValue({ models: ['234'] });
 
-const params = { teamspace: 'ts', federation: '123', project: '123', container: '123' };
+const params = { teamspace: 'ts', federation: '123', project: '123', model: '123' };
```
Comment (carmenfan) [inline backend/tests/v5/unit/middleware/dataConverter/inputs/teamspaces/projects/models/commons/modelSettings.test.js:71]: does that make sense? I don't quite understand why the change here 😕

File: backend/tests/v5/unit/middleware/dataConverter/inputs/teamspaces/projects/models/commons/modelSettings.test.js [hunk 4]
```diff
@@ -53,62 +55,71 @@ Legends.checkLegendExists.mockImplementation((teamspace, model, legend) => {
...
-		[{ params, body: { defaultLegend: 123 } }, false, 'with invalid defaultLegend'],
-		[{ params, body: { defaultLegend: null } }, true, 'with null defaultLegend'],
-		[{ params, body: { defaultLegend: '9c7a6c50-ee85-11e8-af42-09344c707317' } }, false, 'with defaultLegend that does not exist'],
-		[{ params, body: { defaultLegend: '9c7a6c50-ee85-11e8-af42-09344c707317' } }, false, 'with defaultLegend that does not exist'],
-		[{ params, body: { defaultLegend: existingLegendID } }, true, 'with defaultLegend that exists'],
-		[{ params, body: { defaultLegend: existingLegendID } }, true, 'with defaultLegend that exists'],
-		[{ body: {} }, false, 'with empty params, body'],
-		[{}, false, 'with undefined params, body'],
-	])('Check if req arguments for settings update are valid', (data, shouldPass, desc) => {
+		[{ params, body: { name: null } }, MODEL_TYPES.CONTAINER, false, 'with null name'],
```
Comment (carmenfan) [inline backend/tests/v5/unit/middleware/dataConverter/inputs/teamspaces/projects/models/commons/modelSettings.test.js:74]: I think at this point it'll be cleaner if we separate the tests into 3 test groups

File: backend/tests/v5/unit/middleware/dataConverter/inputs/teamspaces/projects/models/commons/modelSettings.test.js [hunk 5]
```diff
@@ -131,8 +142,8 @@ const testValidateAddModelData = () => {
 				type: 'Structure',
 				surveyPoints: [],
 				angleFromNorth: 10,
-			} }, true, 'with a valid request params, body'],
-		[{ params, body: {} }, false, 'if request params, body is empty'],
+			} }, MODEL_TYPES.CONTAINER, true, 'with a valid request params, body'],
```
Comment (carmenfan) [inline backend/tests/v5/unit/middleware/dataConverter/inputs/teamspaces/projects/models/commons/modelSettings.test.js:145]: similar here

File: backend/tests/v5/unit/middleware/dataConverter/outputs/teamspaces/projects/models/commons/modelSettings.test.js [hunk 6]
```diff
@@ -15,6 +15,7 @@
  *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+const { MODEL_TYPES } = require('../../../../../../../../../../src/v5/models/modelSettings.constants');
```
Comment (carmenfan) [inline backend/tests/v5/unit/middleware/dataConverter/outputs/teamspaces/projects/models/commons/modelSettings.test.js:18]: use the src variable (i'm losing track how many times I've had to point this out 🙈 )

**Review by carmenfan — CHANGES_REQUESTED** [review]
- I think the chat/event manager needs looking at, for e.g This is what I'm getting according to the logs when I create a new drawing:
image

Please make sure we're also adding tests for the notifications we expect.
We're also missing additional documentation around the new events in the README.

- documentation needs updating as discussed.
image

- Get settings documentation has not been updated
image

File: backend/tests/v5/unit/middleware/dataConverter/outputs/teamspaces/projects/models/commons/modelSettings.test.js (hunk 6 above)
Comment (ChristosTsiotsias) [reply backend/tests/v5/unit/middleware/dataConverter/outputs/teamspaces/projects/models/commons/modelSettings.test.js:18]: Its auto generated and I keep forgetting to change it 😭

File: backend/src/v5/routes/teamspaces/projects/models/common/general.js [hunk 7]
```diff
@@ -175,11 +165,20 @@ const updateModelSettings = (modelType) => async (req, res) => {
...
+	};
 	try {
-		const { getSettings: fn } = modelType === MODEL_TYPES.federation ? Federations : Containers;
-		const settings = await fn(teamspace, model);
-		req.outputData = settings;
-		await next();
+		const settings = await fn[modelType](teamspace, model);
+
+		if (modelType === MODEL_TYPES.DRAWING) {
+			respond(req, res, templates.ok, settings);
```
[the remark this answers is not in the document]
Comment (ChristosTsiotsias) [reply backend/src/v5/routes/teamspaces/projects/models/common/general.js:177]: If you mean the model Id they are all stored as strings in the db

File: backend/tests/v5/e2e/routes/teamspaces/projects/models/common/general.test.js [hunk 8]
```diff
@@ -65,13 +67,17 @@ const setupBasicData = async (users, teamspace, project, models) => {
...
 	describe('Get model list', () => {
-		const { users, teamspace, project, con, fed } = generateBasicData();
-		const models = times(10, (n) => ({
+		const { users, teamspace, project, con, fed, draw } = generateBasicData();
+		const models = [...times(10, (n) => ({
 			...ServiceHelper.generateRandomModel({ isFederation: n % 2 === 0 }),
 			isFavourite: n % 3 === 0,
-		}));
+		})),
+		...times(10, (n) => ({
```
[the remark this answers is not in the document]
Comment (ChristosTsiotsias) [reply backend/tests/v5/e2e/routes/teamspaces/projects/models/common/general.test.js:75]: I thought about it but then I noticed that we do isFavourite: n % 3 === 0 so only drawing models would be favourites so i decided to make another collection... should I stick with one ? 🤔

File: backend/src/v5/routes/teamspaces/projects/models/common/general.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/src/v5/routes/teamspaces/projects/models/common/general.js:177]: right, could we pass it through the middleware anyway?

File: backend/tests/v5/e2e/routes/teamspaces/projects/models/common/general.test.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/tests/v5/e2e/routes/teamspaces/projects/models/common/general.test.js]: I would suggest you stick with and adjust the logic to suit

**Review by carmenfan — CHANGES_REQUESTED** [review]
- swagger doc for favourites is still saying `model` when it expects con/fed/drawing on favourites endpoint:
image

- Error message is quite cryptic when you plugin a unknown drawing id on the favourites endpoint:
image

**Review by carmenfan — APPROVED** [review]
